### PR TITLE
feat(orders): add NIP-17 order read integration helper

### LINKED ISSUE #1084: [BUG] Message not viewable to sender after sending follow up message after an order to the seller
**Describe the bug**
After making an order, I do not see the messages to the seller in other apps. I originally thought that this was due to using NIP-17 gift wraps that weren't supported by other clients, but I don't see it in the Lightning Piggy mobile app which does support NIP-17. Looking at the logs, I then wondered if you are sending the message via the preferred relays of the sender and/or receiver or just the plebeian relay?

**To Reproduce**

1. Go to buy the following sticker pack: https://plebeian.market/products/4866bd0f50b6828b2b7e3ef048789f8ff59b9daae55b8539de5769e8f9b3ceef
2. Proceed to place the order (using Amber and WebLN to make the payment)
4. Confirm if the recipient received the message

**Expected behavior**
I should be able to see the order messages in other messaging clients.

**Screenshots**

Image

- Browser: Edge
- OS: Kubuntu

**Additional context**
- I signed each event using Amber.
- I used WebLN to make the payment (successfully)
- I sent a direct message using Primal (NIP-04) and Lightning Piggy (NIP-17) and can see those two messages, but not the ORder Placed or Payment Receipt.

### PR DESCRIPTION
**Pull request by maximotodev**
## What

Adds a helper-only ADR-014 read integration boundary for merging legacy public order-message events with already-unwrapped NIP-17 order messages.

## Why

ADR-014’s next migration step is narrow read integration before any checkout, query, UI, publish, or cutover wiring.

This helper keeps the migration staged by returning protocol/domain records only. It does not fetch, decrypt, unwrap, sign, publish, call NDK, call Applesauce, touch React Query, or return `NDKEvent`.

## Scope

- Adds `src/lib/orders/nip17OrderReadIntegration.ts`
- Adds `src/lib/__tests__/nip17OrderReadIntegration.test.ts`
- No `src/queries/orders.tsx`
- No checkout UI
- No React Query changes
- No live relay read/write wiring
- No legacy raw read/write removal
- No storage/cache behavior
- No direct `@nostr-dev-kit/ndk` imports
- No payment settlement or fulfillment-state derivation

## Boundary behavior

- Accepts verified legacy public kind 14, 16, and 17 order-message events
- Validates legacy signed/raw events separately from unsigned NIP-17 rumors
- Accepts already-unwrapped NIP-17 order messages and validates their inner rumors through `assertOrderMessageRumor`
- Revalidates canonical rumor shape and transport metadata defensively
- Uses inner `rumor.tags` only, never gift-wrap or seal tags
- Rejects non-order rumors, malformed metadata, self-addressed messages, and ambiguous multi-recipient messages
- Treats exactly one `p` recipient as the Gamma buyer/merchant order boundary, not as a general NIP-17 limitation
- Accepts kind 14 messages without the optional `subject` tag
- Preserves compatibility with legacy payment-request `recipient` tags and fixed-two-decimal order amounts
- Requires payment-request `amount` and validates known shipping statuses
- Preserves legacy empty payment proofs as read data without deriving `paid` or `settled`
- Dedupes by canonical inner message id, preferring NIP-17 when legacy and NIP-17 ids collide
- Dedupes sender/recipient wrappers and relay duplicates by inner rumor id
- Preserves legacy-only records and sorts deterministically by `created_at`, then `id`

## Validation

Local validation after synchronizing with `master`:

- Focused integration helper: 18 pass, 0 fail
- Related NIP-17/order-message suite: 34 pass, 0 fail
- Full unit suite: 204 pass, 0 fail across 19 files
- Prettier: passed
- `git diff --check` on the PR delta: passed
- NDK footprint: unchanged at 127 / 127

GitHub Actions at `be86f375edde113a185e779a9b82684ce1709fd7`:

- Prettier Check: passed
- NDK Footprint Guard: passed
- Unit and Integration Tests: passed
- E2E pricing job: passed, including pinned `nak` installation, relay/app startup, and pricing tests

## Non-goal

This PR does not wire the helper into live relay queries, checkout, or UI state, and therefore does not by itself close #1084 or prove end-to-end sender visibility.

Refs #1028. Related to #1084.

### CODE REVIEW

**Comment by maximotodev** [comment]
@codex review

File: src/lib/orders/nip17OrderReadIntegration.ts [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+		}
+
+		if (event.kind === ORDER_PROCESS_KIND) {
+			const messageType = event.tags.find((tag) => tag[0] === 'type')?.[1]
+			if (messageType === ORDER_MESSAGE_TYPE.ORDER_CREATION) {
+				OrderCreationSchema.parse(schemaInput)
+				return true
+			}
+			if (messageType === ORDER_MESSAGE_TYPE.PAYMENT_REQUEST) {
+				PaymentRequestSchema.parse(schemaInput)
```
[the remark this answers is not in the document]
Comment (maximotodev) [reply src/lib/orders/nip17OrderReadIntegration.ts:220]: Fixed in the latest push. Legacy kind 16/type 2 payment-request validation now tolerates the existing single well-formed `recipient` tag for compatibility with current checkout output, while preserving the original tag in normalized records. NIP-17 inner rumors remain strict, malformed or duplicate `recipient` tags are rejected, arbitrary extra legacy tags are still rejected, and no direct NDK imports were added.

Validation is green after the amend: footprint, Prettier, unit/integration, security scan, and e2e-pricing all passed.

**Comment by maximotodev** [comment]
@codex review

**Comment by maximotodev** [comment]
@codex review

File: src/lib/orders/nip17OrderReadIntegration.ts [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+	try {
+		if (event.kind === ORDER_GENERAL_KIND) {
+			GeneralCommunicationSchema.parse(schemaInput)
+			return true
+		}
+
+		if (event.kind === ORDER_PROCESS_KIND) {
+			const messageType = event.tags.find((tag) => tag[0] === 'type')?.[1]
+			if (messageType === ORDER_MESSAGE_TYPE.ORDER_CREATION) {
+				OrderCreationSchema.parse(schemaInput)
```
[the remark this answers is not in the document]
Comment (maximotodev) [reply src/lib/orders/nip17OrderReadIntegration.ts:216]: Fixed in the latest push. Legacy kind 16/type 1 order-creation validation now tolerates the existing `.toFixed(2)` whole-decimal amount shape, such as `1000.00`, for compatibility with current checkout output, while preserving the original `amount` tag in normalized records.

NIP-17 inner rumors remain strict, duplicate amount tags and arbitrary extra legacy tags are rejected, and no direct NDK imports were added.

Validation is green locally: focused read-integration test, related helper suite, full unit suite, Prettier, diff check, and NDK footprint guard all passed.

File: src/lib/orders/nip17OrderReadIntegration.ts [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+function isLegacyOrderMessageEvent(value: unknown): value is LegacyOrderMessageEvent {
+	if (!isRecord(value)) return false
+	if (!nonEmptyString(value.id)) return false
+	if (!nonEmptyString(value.pubkey)) return false
+	if (!Number.isInteger(value.created_at)) return false
+	if (!Number.isInteger(value.kind)) return false
+	if (typeof value.content !== 'string') return false
+	if (!isStringTags(value.tags)) return false
+	if ('sig' in value && typeof value.sig !== 'string') return false
+	return true
```
[the remark this answers is not in the document]
Comment (maximotodev) [reply src/lib/orders/nip17OrderReadIntegration.ts:155]: Fixed in the latest push. Legacy raw order-message events now require a valid canonical Nostr event id/signature before normalization.

The existing legacy compatibility shims remain validation-only: kind 16/type 2 payment requests still tolerate one well-formed `recipient` tag, and kind 16/type 1 order creations still tolerate whole-decimal `.toFixed(2)` amounts such as `1000.00`, while preserving the original tags in normalized records.

NIP-17 inner rumors remain strict, no NDK/Applesauce calls were added, and no direct `@nostr-dev-kit/ndk` imports were added.

Validation is green locally: focused read-integration test, related helper suite, full unit suite, Prettier, diff check, and NDK footprint guard all passed.

**Comment by maximotodev** [comment]
@codex review

File: src/lib/orders/nip17OrderReadIntegration.ts [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+		if (!recipientPubkey) return null
+		if (value.recipientPubkey !== recipientPubkey) return null
+
+		if (value.direction === 'sent') {
+			if (value.userPubkey !== rumor.pubkey) return null
+			if (value.counterpartyPubkey !== recipientPubkey) return null
+		} else {
+			if (value.userPubkey !== recipientPubkey) return null
+			if (value.counterpartyPubkey !== rumor.pubkey) return null
+		}
```
[the remark this answers is not in the document]
Comment (maximotodev) [reply src/lib/orders/nip17OrderReadIntegration.ts:129]: Fixed in the latest push. The read integration helper now rejects self-addressed NIP-17 inner rumors where the rumor author pubkey is the same as the single recipient `p` tag, matching the upstream unwrap boundary that requires the active user to be exactly one side of the conversation.

Valid NIP-17 messages still normalize, legacy validation/signature checks remain unchanged, and the existing legacy compatibility shims for kind 16/type 2 `recipient` tags and kind 16/type 1 `.toFixed(2)` whole-decimal amounts remain validation-only.

No NDK/Applesauce calls or direct `@nostr-dev-kit/ndk` imports were added.

Validation is green locally and in CI: focused read-integration test, related helper suite, full unit suite, Prettier, diff check, NDK footprint guard, security scan, unit/integration, and e2e-pricing all passed.

The previous Codex comments remain addressed: payment-request `recipient` compatibility, decimal `.00` legacy order amount compatibility, and legacy signature verification are all still covered by the passing tests.

**Comment by maximotodev** [comment]
@codex review

File: src/lib/orders/nip17OrderReadIntegration.ts [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+function tagsForLegacyOrderCreationSchemaValidation(tags: string[][]): string[][] | null {
+	const amountTags = tags.filter((tag) => tag[0] === 'amount')
+	if (amountTags.length !== 1) return null
+
+	const amountTag = amountTags[0]
+	const amountValue = amountTag?.[1]
+	if (!nonEmptyString(amountValue)) return null
+
+	if (INTEGER_AMOUNT_RE.test(amountValue)) return tags
+	if (!LEGACY_DECIMAL_WHOLE_AMOUNT_RE.test(amountValue)) return tags
```
[the remark this answers is not in the document]
Comment (maximotodev) [reply src/lib/orders/nip17OrderReadIntegration.ts]: Fixed in the latest push. Legacy kind 16/type 1 order-creation validation now tolerates the existing `.toFixed(2)` amount shape, including non-zero decimal values such as `12.50`, for compatibility with current checkout output.

The decimal amount normalization is validation-only; returned records still preserve the original `amount` tag unchanged. Duplicate amount tags, arbitrary extra legacy tags, and malformed decimal shapes are still rejected.

NIP-17 inner rumors remain strict, legacy signature verification remains unchanged, self-addressed NIP-17 rumors remain rejected, and no NDK/Applesauce calls or direct `@nostr-dev-kit/ndk` imports were added.

Validation is green locally and in CI: focused read-integration test, related helper suite, full unit suite, Prettier, diff check, NDK footprint guard, security scan, unit/integration, and e2e-pricing all passed.

**Comment by c03rad0r** [comment]
Items worth carrying over from #1128 (now closed, superseded by this PR):

1. Test coverage: #1128 had 739 lines of tests vs 572 here. Some edge cases tested there are not yet covered here — e.g., same rumor appearing in both sent and received directions, empty payment proof preservation. Worth porting any unique scenarios.

2. Franchovy's 5 review points from #1128 apply here too: subject tag is marked required but NIP spec says optional for kind 14; scope of cryptographic verification on NIP-17 content; shipping status field validation; missing amount tag in payment requests; filtering NIP-17 messages with non-order rumor kinds.

3. AGENTS.md: #1128 added documentation in src/lib/orders/. This PR does not have it yet — worth carrying over.

**Review by c03rad0r — COMMENTED** [review]
Reviewed the full diff. This is well-structured — pure function, clean protocol boundary, good test coverage.

The architectural decision to return OrderMessageReadRecord (no NDK types) is the right one for the migration. This is the correct evolution of #1128's approach.

A few notes:

1. This appears to supersede #1128. I'd suggest we close #1128 and port any unique test cases from there. Can you confirm?

2. Missing AGENTS.md in src/lib/orders/ — #1128 added one. Worth carrying over.

3. The legacy decimal amount shim (.toFixed(2) tolerance) — could you add a test case for a non-zero decimal like 12.50? The tests cover integers and .00 but I didn't see a non-trivial decimal case.

4. The single p tag requirement on NIP-17 rumors — is there a scenario where a legitimate order message could have multiple p tags (e.g., multi-party)? If not, worth a one-line comment explaining why exactly-one is correct.

CI is green. No blockers — just the above items. Happy to approve once clarified.

**Comment by c03rad0r** [comment]
Reviewed the full diff. This is well-structured — pure function, clean protocol boundary, good test coverage.

The architectural decision to return `OrderMessageReadRecord` (no NDK types) is the right one for the migration. This is the correct evolution of #1128's approach.

A few notes:

1. This appears to supersede #1128. I'd suggest we close #1128 and port any unique test cases from there. Can you confirm?

2. Missing `AGENTS.md` in `src/lib/orders/` — #1128 added one. Worth carrying over.

3. The legacy decimal amount shim (`.toFixed(2)` tolerance) — could you add a test case for a non-zero decimal like `12.50`? The tests cover integers and `.00` but I didn't see a non-trivial decimal case.

4. The single `p` tag requirement on NIP-17 rumors — is there a scenario where a legitimate order message could have multiple `p` tags (e.g., multi-party)? If not, worth a one-line comment explaining why exactly-one is correct.

CI is green. No blockers — just the above items. Happy to approve once clarified.

---

**Items worth porting from #1128** (if we close it in favor of this PR):

- **Test coverage:** #1128 has 739 lines of tests vs 572 here. Some test scenarios in #1128 may not be covered yet — worth diffing the test files before closing.
- **Franchovy's 5 review points on #1128** raised valid concerns that apply here too: subject tag optionality for kind 14 (spec says optional, #1128 kept it required with justification), crypto verification scope (kept out of scope since messages arrive pre-unwrapped), shipping status validation, missing-amount test case, non-order rumor-kind filtering test. Worth reviewing whether those are addressed here.
- **AGENTS.md:** #1128 added `src/lib/orders/AGENTS.md`. Should be carried over to this PR.

**Comment by Franchovy** [comment]
Overall looks like a cleanly scoped PR. The only downside is the lack of context brought forward into the new code itself. I would suggest making these improvements. Change Requests are non-blocking:

### Severity: Important
Location: isVerifiedLegacyOrderMessageEvent() (lines ~145-154)
Issue: The PR adds verifyEvent() from nostr-tools, but this function's behavior depends on the relay trust model. According to ADR-013, relay data is untrusted.

While verifyEvent() validates cryptographic signatures, it doesn't validate:

- Whether the pubkey matches the expected sender/counterparty for the order
- Whether the event timestamp is reasonable (could be replay attacks)
- Whether the event ID is canonical (the code already checks this separately)
- Recommendation: Add inline comment acknowledging this limitation:

```
/**
 * Verifies Nostr signature validity. Note: This does NOT validate that the
 * pubkey belongs to the expected order participant. Caller must ensure
 * event origin matches order context (buyer vs seller) before merge.
 */
 ```

### Missing AGENTS.md Documentation
Severity: Non-blocker (but required)
Location: src/lib/orders/AGENTS.md
Issue: Per reviewer c03rad0r's comment, this file was added in #1128 but is missing from this PR. This is a documentation regression.

Action Required: Copy AGENTS.md from #1128 or create equivalent documentation explaining:

- Purpose of this helper
- How to extend the normalization logic
- Testing conventions
- Relationship to ADR-013 and ADR-014


### Transport Metadata Optional Fields
Severity: Nitpick
Location: OrderMessageReadRecord type (lines ~35-44)
Issue: The transport object has mostly optional fields (rumorId?, giftWrapId?, etc.). However, for NIP-17 records, all these should be present (not optional) since they're populated in `nip17TransportMetadata()`.

Recommendation: Make transport fields non-optional when `source === 'nip17'`:

```
export type OrderMessageReadRecord = {
  source: OrderMessageReadSource
  // ... other fields
  transport?: OrderMessageReadTransport // Currently optional for both sources
}

// Could add:
export type Nip17OrderMessageReadRecord = OrderMessageReadRecord & {
  source: 'nip17'
  transport: NonNullable<OrderMessageReadRecord['transport']>
}
```

This would give better TypeScript guarantees downstream.

**Review by Franchovy — APPROVED** [review]
Non-blocking changes requested

**Comment by maximotodev** [comment]
Thanks for the review! Addressed in `49a5d27f`. I added the authentication-versus-order-context comment, introduced a discriminated union requiring transport metadata for NIP-17 records while prohibiting it for legacy records, and added narrowing coverage. `giftWrapId` remains optional because this defensive boundary may legitimately lack a wrapper ID. I also verified that #1128 did not add `src/lib/orders/AGENTS.md`; the existing parent `src/lib/AGENTS.md` is already present, so I avoided duplicating stale documentation. Local validation is green: Prettier, 18 focused tests, 34 related tests, 204 unit tests, and NDK footprint `127/127`.

**Comment by maxime-tt** [comment]
## Review — feat(orders): add NIP-17 order read integration helper

Reviewed against ADR-013, ADR-014, and `src/lib/AGENTS.md`. Scope, privacy, and payment-lifecycle discipline all look correct. Verified locally: 18/18 tests pass (112 expects); zero new `@nostr-dev-kit` runtime imports in the new files; NDK footprint unchanged.

### 🟠 should-fix / clarify

**1. Fixed-two-decimal `amount` compatibility is asymmetric across message kinds**
`src/lib/orders/nip17OrderReadIntegration.ts` `tagsForLegacyOrderCreationSchemaValidation` only relaxes the integer `amount` regex for `ORDER_CREATION`. `PAYMENT_REQUEST` and `PAYMENT_RECEIPT` share the same `AmountTagSchema` (`/^\d+$/`, `src/lib/schemas/order.ts`) but get no decimal normalization, so a legacy payment request/receipt carrying `"2100.00"` fails schema validation and is **silently dropped** from the merged read set. The PR description says "fixed-two-decimal order amounts" without scoping it to order-creation only.
*Why it matters:* silent drops of legitimate legacy messages break read fidelity for existing orders during the migration window — ADR-014 Decision 7 requires legacy raw reads to remain.
*Suggested fix:* either (a) confirm legacy never used decimal amounts for requests/receipts and add a test asserting they are rejected, or (b) extend the same normalization to those kinds. Either way, narrow the wording in the PR/`tagsForLegacySchemaValidation`.

### 🟡 nits

**2. Decimal replacement is semantically misleading**
`amountValue.replace('.', '')` turns `"1000.00"` into `"100000"` purely to satisfy the integer regex; the canonical value passed downstream is the unchanged original `"1000.00"`. A reader may assume it converts to integer sats (`"1000"`), which it does not. Prefer `String(Math.round(Number(amountValue)))` or a comment clarifying that only the regex is satisfied and the canonical string is preserved.

**3. DRY: recipient-extraction duplicated**
`getOrderRumorRecipientPubkey` exists in both `nip17OrderRead.ts` (throws) and here (returns `null`) with the same two-party `p`-tag logic. Consider a shared helper returning a discriminated result so both callers reuse one implementation; the different failure modes justify the split today but it's a drift risk as the boundary evolves.

**4. Legacy author authorization is deferred (track it)**
`isVerifiedLegacyOrderMessageEvent` authenticates the signature only and does not confirm the author is a participant in the specific order — the code comment states this honestly. Because legacy kind 14/16/17 events are public and unencrypted, a relay can inject a correctly-signed unrelated message with the user as the `p` recipient and this helper will surface it. Not exploitable now (helper is unwired), but the eventual consumer in `src/queries/orders` MUST apply order-context filtering before trusting authorship. Worth a tracked TODO so it isn't lost at wiring time.

### Positives / verification
- **Scope:** only the two files added; no queries/checkout/UI/relay-IO wiring; matches ADR-014 Decision 2 & 9 (narrow seam, no `src/queries/orders.tsx` rewrite) and the PR's stated purpose.
- **NDK boundary:** no new `@nostr-dev-kit` runtime imports — helper uses `nostr-tools` `verifyEvent` plus a type-only import of `UnwrappedNip17OrderMessage`. Consistent with `src/AGENTS.md` (ADR-0002) and the NDK footprint guard. Confirmed: 0 NDK imports in the new files.
- **Privacy (ADR-014 D10/D12):** only inner `rumor.tags` are used; wrapper/seal tags are never surfaced (test "does not expose wrapper or gift-wrap tags"). No logging of decrypted content/proofs.
- **Payment lifecycle (`src/lib/AGENTS.md`):** no `paid`/`settled` derivation; empty-proof receipts preserved as read data (test confirms). Correct.
- **Authenticity:** legacy events verified via `verifyEvent` (recomputes id + sig, so `id` is authenticated); NIP-17 rumors revalidated via `assertOrderMessageRumor` (canonical id), with self-addressed, multi-recipient, recipient-mismatch, and direction-inconsistency all rejected defensively.
- **Determinism:** sort by `created_at` then `id.localeCompare`; NIP-17 preferred over legacy on canonical-id collision; dedupe by inner rumor id (ADR-014 D6). Verified by tests.

### Verdict
**APPROVE with nits.** Recommend resolving #1 (clarify or extend decimal-amount coverage) before this is wired into live reads; #2–#4 are reasonable follow-ups. The helper is a clean, well-tested, read-only boundary that respects the migration staging in ADR-014.

**Review by c03rad0r — APPROVED** [review]

**Review by Franchovy — APPROVED** [review]
